src/main/java/groovy/lang/GString.java

```diff
@@ -101,9 +101,12 @@ public Object invokeMethod(String name, Object args) {
     }
 
     public GString plus(GString that) {
-        Object[] values = getValues();
+        Object[] v1 = this instanceof GStringImpl ? ((GStringImpl)this).getValueList().toArray(new Object[0]) : getValues();
```

Comment:
We could use `EMPTY_STRING_ARRAY` and `EMPTY_OBJECT_ARRAY` to avoid creating arrays repeatly

Reply:
`this instanceof GStringImpl ? ((GStringImpl)this).getValueList().toArray(new Object[0]) : getValues()` and `this instanceof GStringImpl ? ((GStringImpl)this).getStringList().toArray(new String[0]) : getStrings()` could be extracted to seperate methods